Fix Python3 binding's submodules

adamgreig opened the pull request:
Previously the Python3 `cv2` package ends up with no submodules (`bgsegm`, `face`, etc) in it, which makes a lot of functionality unusable. By not writing over our `root` reference we ensure the new submodules are added to the correct `cv2` module.

I'm not entirely clear why this is necessary (and in Python 2 it works fine either way) but my suspicion is that somewhere (likely the second run through the `while` loop) an unsuitable reference to `cv2` is being created. I initially ran into this issue with the opencv_contrib modules but it affects all the submodule namespaces.

Ideally some of the tests would catch this but it looks like they all use `env Python`, are any run under Python 3?

Before change, Python 3:

````
>>> import cv2
>>> cv2.xfeatures2d.SIFT_create()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'xfeatures2d'
````

Afterwards:

````
>>> import cv2
>>> cv2.xfeatures2d.SIFT_create()
<xfeatures2d_SIFT 0x7f6942a83f50>
```

adamgreig commented:
I suspect this is the issue behind these questions on SE too:

http://stackoverflow.com/questions/28484893/cannot-import-xfeatures2d-in-python-opencv-binding

http://stackoverflow.com/questions/26938816/opencv-3-0-0-beta-missing-xfeatures2d-submodule-python3-4

http://stackoverflow.com/questions/26059134/adding-modules-from-opencv-contrib-to-opencv

vpisarev commented:
@mshabunin, can you please review it?

mshabunin commented:
@adamgreig, thank you for contribution, it works for me :+1:
Concerning python tests, we currently only build _python2_ library and run only 6 tests on `master` branch: [example](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/2680/steps/test_python/logs/stdio)

adamgreig commented:
Thanks for looking! Is it simple to add python3 building and tests?

I've just noticed that I think the fix as I've got here would break anything that actually used multi-level paths, like "cv2.bla.bla.something", which nothing does at the moment (hence it still working fine). The problem would be when we run `PyObject * d = PyModule_GetDict(root);` and it always points at the original root rather than the new submodule. But setting this to submod means we end up with the old (broken) behaviour.

I have an idea on how to fix that so I'll try it out this weekend.

adamgreig commented:
@mshabunin ok, that commit should fix the behaviour in the nested submodule case (though there aren't any defined at the moment so it's a bit hard to test :p). It still works in Python 2 and 3.

mshabunin commented:
@adamgreig, the python3 does not work for me now: there is no submodules. The python2 works well, submodules (with nested) are created. Can you, please, check it again?

adamgreig commented:
@mshabunin after a little more thought and testing, the problem seems to be that in Py3, the `PyDict_GetItemString` when called the first time (with short_name as "") returns a copy of the cv2 module at a new memory address, so when that gets updated it's not reflected in the module that was passed in as root. Not really sure why still. Doesn't appear to happen in Py2.

Anyway with this much smaller fix submodules (nested or otherwise) work fine on Py2 and Py3 now. I've squashed the commits down to just this one and rebased off current master.

mshabunin commented:
@adamgreig, thank you, now it works in both versions for me. 
Also, I've added _python3_ test run during pull request check ([for linux builds](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/2735/steps/test_python3/logs/stdio)).
:+1:
